TapeAgents export for experiment results

ollmer opened the pull request:
This PR introduces a simple compatibility layer for exporting traces of the BrowserGym experiments in a format compatible with the Tapes of the [TapeAgents framework](https://github.com/ServiceNow/TapeAgents).
What does it do:
- convert a single trace of the experiment into a single Tape
- decomposes each StepInfo step of the experiment into 3 steps of the tape agents: observation, thought, and action.
- attempts to preserve as much metadata as possible for the step and for the whole trace.
- parses action in the simplest possible way as a string with a single Python function call
- if the `think` field is present in the agent info and not empty - it creates the thought step of the tape from it.
- method to save the tape in json format

The MVP is ready but not yet tested.

Next steps in this PR: 
- test conversion over the real traces
- collect feedback from the BrowserGym team
- test tape loading on the TapeAgents side and rendering it with the tape browser

cc @recursix.

recursix commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
+                screenshot: str = ""
+                screenshot_som: str = ""
+                if "screenshot" in step_info.obs:
+                    screenshot = f"{self.exp_dir}/screenshot_step_{step_info.step}.png"
```
leverage pathlib

```suggestion
                    screenshot = {self.exp_dir} / f"screenshot_step_{step_info.step}.png"
```

recursix commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
+                screenshot_som: str = ""
+                if "screenshot" in step_info.obs:
+                    screenshot = f"{self.exp_dir}/screenshot_step_{step_info.step}.png"
+                    step_info.obs.pop("screenshot")
```
this is also popping in the stored self.steps_info, which could lead to unexpected errors if someon tries to use results after calling get_tape

recursix commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
+
+            # extract thought step
+            think = ""
+            if isinstance(step_info.agent_info, AgentInfo):
```
AgentInfo also behaves like a dict, so you could simplify by just doing .get

recursix commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +658,75 @@ def summary_info(self) -> dict:
+                )
+                tape_steps.append(observation)
+
+            # extract thought step
```
We should probably look if there is a tape_sgement first, like we discussed. Or should we wait until it's implemented?

recursix replied:
Could a tape_segment simply be a Tape that can be appened

ollmer replied:
We could merge this implementation to support existing traces and work on the segments separately. I think that segments would be either Tape or list[Step], cause trace metadata is not really needed in a segment

ollmer replied:
I've added support for tape_segment dicts stored in the `other_info` dict of the agent info.

ollmer commented:
I've simplified this PR down to <100 lines of changes in a single file and added support of tape_segments inside the agent_info, which could be produced by TapeAgent in the AgentLab.

gasse commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
+            steps += tape_segment
+        metadata = dict(
+            id=str(uuid.uuid4()),
+            author="browsergym_agent",
```
Maybe the agent name could be re-used here?
```suggestion
            author=f"browsergym_agent_[{agent_name}]",
```

gasse commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
+        if step_info.obs is not None:
+            screenshot: str = ""
+            screenshot_som: str = ""
+            obs_dict = step_info.obs.copy()
```
Should this be a deepcopy?
Drawback: it's costly
Pro: tapes can safely be altered afterwards

ollmer replied:
Yep, it's a good idea. I assumed earlier that the observation fields are mostly primitive types and will be copied anyway, but it's safer to make a deep copy. As the export should rarely be called, complexity is not an issue here. I'll update it, thanks!

gasse commented on browsergym/experiments/src/browsergym/experiments/loop.py:
```diff
@@ -657,6 +657,94 @@ def summary_info(self) -> dict:
+
+        # extract action step
+        action_str = step_info.action
+        # TODO: this is a naive way to split the arguments, make proper parsing later
```
We have a parser in browsergym to extract function calls from raw text actions, you could just reuse it.

There is one caveat though, actions in browsergym can be arbitrary strings (python code, function calls, json). Even with our `function call` action set, agents can be given the freedom to execute several function calls in the same action (e.g. `type("a15", "Hello") click("a16")`).

Can TapeAgent handle arbitrary actions? Do you need the `name` and `arguments` key?

gasse replied:
You'll find an example here:
https://github.com/ServiceNow/BrowserGym/blob/80e063bca66e766c3ec3f3ea2913127dc1bb4b69/tests/core/test_actions_highlevel.py#L102

gasse replied:
The parser itself can be imported from here
https://github.com/ServiceNow/BrowserGym/blob/80e063bca66e766c3ec3f3ea2913127dc1bb4b69/browsergym/core/src/browsergym/core/action/parsers.py#L85

ollmer replied:
I've implemented it, please take a look when you have time.

gasse reviewed (commented):
Just made a few comments. Feel free to ping me for a chat @ollmer

ollmer commented:
Thank you!
